src/qibo/tensorflow/cgates.py

```diff
@@ -737,10 +737,12 @@ def _prepare(self):
             inv_gate = self._invert(gate)
             # use a ``set`` for this loop because it may be ``inv_gate == gate``
             for g in {gate, inv_gate}:
-                g.density_matrix = self.density_matrix
-                g.device = self.device
-                g.nqubits = self.nqubits
+                if g is not None:
```

Comment:
Is it possible to have `None` here?

Reply:
Yes, this may be a bit obscure as it is implemented now. I think the only case it happens is when the gate is the Collapse gate which has no inverse, so I use `None` for the inverse gate (see the `_invert` method definition in `ResetChannel` - line 818 of `cgates.py`). The Collapse gate is non-invertible mathematically speaking anyway, since it is not really a "gate" but the result of measuring. Both `ResetChannel` and `ThermalRelaxationChannel` use the `Collapse` gate.

By the way, it is generally not required to invert gates in order to apply the channels. The reason I do this here is that custom operators work with in-place updates, while when applying channels we need to apply several gates to the same state and then sum the terms. There are two ways to do this:
1. Create a copy of the state (can be done for example with a trick `temp_state = state + tf.zeros_like(state)`) and use this in the custom operators, or
2. Apply each channel gate to the state but then apply its inverse to reset to to the original state before applying the next channel gate (what I do now).

The second method is a bit more complicated in terms of code but I used it because it was slightly faster in some tests I did.